pyro/infer/mcmc/logger.py

```diff
@@ -0,0 +1,111 @@
+import json
+import logging
+import sys
+from collections import OrderedDict
+
+from tqdm.autonotebook import tqdm
+
+LOG_MSG = "LOG"
+TQDM_MSG = "TQDM"
+
+
+class TqdmHandler(logging.StreamHandler):
+    """
+    Handler that synchronizes the log output with the
+    :class:`~tqdm.tqdm` progress bar.
+    """
+    def emit(self, record):
+        try:
+            msg = self.format(record)
+            self.flush()
+            tqdm.write(msg, file=sys.stderr)
+        except (KeyboardInterrupt, SystemExit):
+            raise
+        except Exception:
+            self.handleError(record)
+
+
+class MCMCLoggingHandler(logging.Handler):
+    """
+    Main logging handler used by :class:`~pyro.infer.mcmc`,
+    to handle both progress bar updates and regular `logging`
+    messages.
+
+    :param log_handler: default log handler for logging
+        output.
+    :param progress_bar: If provided, diagnostic information
+        is updated using the bar.
+    """
+    def __init__(self, log_handler, progress_bar=None):
+        logging.Handler.__init__(self)
+        self.log_handler = log_handler
+        self.progress_bar = progress_bar
+
+    def emit(self, record):
+        try:
+            if self.progress_bar and record.msg_type == TQDM_MSG:
+                diagnostics = json.loads(record.getMessage(),
+                                         object_pairs_hook=OrderedDict)
+                self.progress_bar.set_postfix(diagnostics)
+                self.progress_bar.update()
+            else:
+                self.log_handler.handle(record)
+        except (KeyboardInterrupt, SystemExit):
+            raise
+        except Exception:
+            self.handleError(record)
+
+
+class MetadataFilter(logging.Filter):
+    """
+    Adds auxiliary information to log records, like `chain_id` and
+    `msg_type`.
```

Comment:
Adding this auxiliary information (`chain_id`, `msg_type`) will be helpful in the multiprocessing case, where there will be a separate listener thread in the main process, that will listen for logging events and log them appropriately.